# Практическое задание

Тема: Системы линейных уравнений, часть 1

## Задание 1

Решить систему уравнений методом Гаусса:

$$\begin{cases}
x_{1}+x_{2}-x_{3}-2x_{4}=0, \\
2x_{1}+x_{2}-x_{3}+x_{4}=-2, \\
x_{1}+x_{2}-3x_{3}+x_{4}=4.
\end{cases}$$

Составим расширенную матрицу и приведем ее к диагональному виду

$
\begin{pmatrix}
1 & 1 & -1 & -2 & 0 \\
2 & 1 & -1 & 1 & -2 \\
1 & 1 & -3 & 1 & 4
\end{pmatrix}
$

вычитаем из второй строчки удвоенную первую

$
\begin{pmatrix}
1 & 1 & -1 & -2 & 0 \\
0 & -1 & 1 & 5 & -2 \\
1 & 1 & -3 & 1 & 4
\end{pmatrix}
$

вычитаем из третьей строки первую

$
\begin{pmatrix}
1 & 1 & -1 & -2 & 0 \\
0 & -1 & 1 & 5 & -2 \\
0 & 0 & -2 & 3 & 4
\end{pmatrix}
$



Ранг матрицы = 3 а размерность 4, следовательно получим бесконечное множество решений

Перепишем новую систему уравнений


$
\begin{cases}
x_1 + x_2 - x_3 - 2x_4 = 0 \\
-x_2 + x_3 + 5x_4 = -2 \\
-2x_3 + 3x_4 = 4
\end{cases}
$

Примем за параметр $x_2 = c$

$
\begin{cases}
x_1 + c - x_3 - 2x_4 = 0 \\
-c + x_3 + 5x_4 = -2 \\
-2x_3 + 3x_4 = 4
\end{cases}\implies
\begin{cases}
x_1 + c - 2x_4 = x_3 \\
-c + x_3 + 5x_4 = -2 \\
x_3 = \frac{3x_4 - 4}{2}
\end{cases}\implies
\begin{cases}
x_1 + c - 2x_4 = x_3 \\
-c + x_3 + 5x_4 = -2 \\
x_3 = \frac{3x_4 - 4}{2}
\end{cases}\implies
\begin{cases}
2x_1 + 2c - 4x_4 = 3x_4 - 4 \\
-c + x_3 + 5x_4 = -2 \\
x_3 = \frac{3x_4 - 4}{2}
\end{cases}\implies
\begin{cases}
x_1 = \frac{7x_4 - 4 - 2c}{2}\\
x_4 = \frac{2c}{13} \\
x_3 = \frac{3x_4 - 4}{2}
\end{cases}\implies
\begin{cases}
x_1 = 6c - 58 \\
x_4 = \frac{2c}{13} \\
x_3 = 3c - 26
\end{cases}
$

## Задание 2

Проверить на совместность и выяснить, сколько решений будет иметь система линейных уравнений:

   а) $\begin{cases}
3x_{1}-x_{2}+x_{3}=4, \\
2x_{1}-5x_{2}-3x_{3}=-17, \\
x_{1}+x_{2}-x_{3}=0;
\end{cases}$
    
   б) $\begin{cases}
2x_{1}-4x_{2}+6x_{3}=1, \\
x_{1}-2x_{2}+3x_{3}=-2, \\
3x_{1}-6x_{2}+9x_{3}=5;
\end{cases}$
    
   в) $\begin{cases}
x_{1}+2x_{2}+5x_{3}=4, \\
3x_{1}+x_{2}-8x_{3}=-2. 
\end{cases}$

a) Запишем матрицу системы
$
\begin{pmatrix}
3 & -1 & 1 & 4 \\
2 & -5 & -3 & -17 \\
1 & 1 & -1 & 0
\end{pmatrix}\implies
\begin{pmatrix}
1 & 1 & -1 & 0 \\
0 & -4 & -2 & 4 \\
0 & -7 & -1 & -17
\end{pmatrix}\implies
\begin{pmatrix}
1 & 1 & -1 & 0 \\
0 & -4 & -2 & 4 \\
0 & 0 & 2.5 & -24
\end{pmatrix}\implies
$ совместна, 1 решение

б) Запишем матрицу системы
$
\begin{pmatrix}
2 & -4 & 6 & 1 \\
1 & -2 & 3 & -2 \\
3 & -6 & 9 & 5
\end{pmatrix}\implies
\begin{pmatrix}
1 & -2 & 3 & -2 \\
2 & -4 & 6 & 1 \\
3 & -6 & 9 & 5
\end{pmatrix}\implies
\begin{pmatrix}
1 & -2 & 3 & -2 \\
0 & 0 & 0 & 5 \\
3 & -6 & 9 & 5
\end{pmatrix}\implies
0 \neq 5 \implies
$ несовместна

в) Запишем матрицу системы
$
\begin{pmatrix}
1 & 2 & 5 & 4 \\
3 & 1 & -8 & -2 
\end{pmatrix}\implies
\begin{pmatrix}
1 & 2 & 5 & 4 \\
0 & -5 & 7 & -14 
\end{pmatrix}\implies
$
r = 2, n = 4, совместна, бесконечное множество решений

## Задание 3

Проверить на совместность и выяснить, сколько решений будет иметь система линейных уравнений, заданная расширенной матрицей

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 3 & -2 & 4\\ 
0 & 5 & 0 & 1\\ 
0 & 0 & 3 & 0\\ 
0 & 0 & 0 & 2
\end{matrix}\right|
\begin{matrix}
3\\ 
2\\
4\\
1
\end{matrix}
\end{pmatrix}.$$

In [ ]:
import numpy as np
A = np.array([[1, 3, -2, 4], [0, 5, 0, 1], [0, 0, 3, 0], [0, 0, 0, 2]])
np.linalg.det(A)

30.000000000000004

Так как определитель не равен 0, система может иметь бесконечное множество корней, но можем найти и конкретное решение, если подобрать обратную матрицу для матрицы A

In [ ]:
A_inv = np.linalg.inv(A)
b = np.array([3, 2, 4, 1])

In [ ]:
np.dot(A_inv, b)

array([2.76666667, 0.3       , 1.33333333, 0.5       ])

Система имеет конкретное решение $[2.76666667, 0.3, 1\frac{1}{3}, 0.5]$

## Задание

Дана система линейных уравнений, заданная расширенной матрицей

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{matrix}\right|
\begin{matrix}
a\\ 
b\\
c
\end{matrix}
\end{pmatrix}.$$

Найти соотношение между параметрами $a$, $b$ и $c$, при которых система является несовместной.

$A^{-1} = \frac{1}{|A|}A_*^T$

In [ ]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
det_A = np.linalg.det(A)
det_A

6.66133814775094e-16

Определитель равен нулю, следовательно обратной матрицы $A^{-1}$ не существует, это означает: что система несовместна при любых a, b, c

# Практическое задание

Тема: Системы линейных уравнений, часть 2

## Задание 1

Решить систему уравнений методом Крамера:

   а) $\begin{cases}
x_{1}-2x_{2}=1 \\
3x_{1}-4x_{2}=7
\end{cases}$
    
   б) $\begin{cases}
2x_{1}-x_{2}+5x_{3}=10 \\
x_{1}+x_{2}-3x_{3}=-2 \\
2x_{1}+4x_{2}+x_{3}=1
\end{cases}$

$$x_{i}=\frac{detA_{i}}{detA}$$

### Решение

a)
$A =
\begin{pmatrix}
1 & -2 \\
3 & -4
\end{pmatrix}$

$|A| = -4 + 6 = 2$

$
A_{1} =
\begin{pmatrix}
1 & -2 \\
7 & -4
\end{pmatrix} = -4 + 14 = 10
$

$A_{2} =
\begin{pmatrix}
1 & 3 \\
1 & 7
\end{pmatrix} = 7 - 3 = 4
$

$x_1 = 5, x_2 = 2$

b)
$A =
\begin{pmatrix}
2 & -1 & 5 \\
1 & 1 & -3 \\
2 & 4 & 1
\end{pmatrix},
B = \begin{pmatrix}
10 \\
-2 \\
1
\end{pmatrix}
$

$|A|=2
\begin{pmatrix}
1 & -3 \\
4 & 1
\end{pmatrix}
+ 1
\begin{pmatrix}
1 & -3 \\
2 & 1
\end{pmatrix}
+ 5
\begin{pmatrix}
1 & 1 \\
2 & 4
\end{pmatrix}= 26 + 7 + 10 = 43$
$

$
A_1 = 10\begin{pmatrix}
1 & -3 \\
4 & 1
\end{pmatrix} + 1\begin{pmatrix}
-2 & -3 \\
1 & 1
\end{pmatrix} + 5\begin{pmatrix}
-2 & 1 \\
1 & 4
\end{pmatrix}
= 86
$

$
A_2 = 2\begin{pmatrix}
-2 & -3 \\
1 & 1
\end{pmatrix} - 10\begin{pmatrix}
1 & -3 \\
2 & 1
\end{pmatrix} + 5\begin{pmatrix}
1 & -2 \\
2 & 1
\end{pmatrix}
= -43
$

$
A_3 = 2\begin{pmatrix}
1 & -2 \\
4 & 1
\end{pmatrix} + 1\begin{pmatrix}
1 & -2 \\
2 & 1
\end{pmatrix} + 10\begin{pmatrix}
1 & 1 \\
2 & 4
\end{pmatrix}
= 43
$

$x_1 = 2, x_2 = -1, x_3 = 1$

## Задание 2

Найти $L$-матрицу $LU$-разложения для матрицы коэффициентов:

   а)$$\begin{pmatrix}
1 & 2 & 4 \\ 
2 & 9 & 12 \\ 
3 & 26 & 30
\end{pmatrix}$$
    
   б)$$\begin{pmatrix}
1 & 1 & 2 & 4\\ 
2 & 5 & 8 & 9\\ 
3 & 18 & 29 & 18\\
4 & 22 & 53 & 33
\end{pmatrix}$$

### Решение

a) $\begin{pmatrix}
1 & 2 & 4 \\ 
2 & 9 & 12 \\ 
3 & 26 & 30
\end{pmatrix}$

1)
$
L = \begin{pmatrix}
1 & 0 & 0 \\
2 & 1 & 0 \\
3 & l_{3, 2} & 1 \\
\end{pmatrix},
U = \begin{pmatrix}
1 & 2 & 4 \\
0 & 5 & 4 \\
0 & 20 & 18
\end{pmatrix}
$

2)
$
L = \begin{pmatrix}
1 & 0 & 0 \\
2 & 1 & 0 \\
3 & 4 & 1 \\
\end{pmatrix},
U = \begin{pmatrix}
1 & 2 & 4 \\
0 & 5 & 4 \\
0 & 0 & 2
\end{pmatrix}
$

b)
$\begin{pmatrix}
1 & 1 & 2 & 4\\ 
2 & 5 & 8 & 9\\ 
3 & 18 & 29 & 18\\
4 & 22 & 53 & 33
\end{pmatrix}$

1) $L = \begin{pmatrix}
1 & 0 & 0 & 0 \\
2 & 1 & 0 & 0 \\
3 & l_{3,2} & 1 & 0 \\
4 & l_{4,2} & l_{4,3} & 1
\end{pmatrix}, U = \begin{pmatrix}
1 & 1 & 2 & 4 \\
0 & 3 & 4 & 1 \\
0 & 15 & 23 & 6 \\
0 & 18 & 45 & 17 \\
\end{pmatrix}$

2) $L = \begin{pmatrix}
1 & 0 & 0 & 0 \\
2 & 1 & 0 & 0 \\
3 & 5 & 1 & 0 \\
4 & 6 & l_{4,3} & 1
\end{pmatrix}, U = \begin{pmatrix}
1 & 1 & 2 & 4 \\
0 & 3 & 4 & 1 \\
0 & 0 & 3 & 1 \\
0 & 0 & 21 & 19
\end{pmatrix}$

3) $L = \begin{pmatrix}
1 & 0 & 0 & 0 \\
2 & 1 & 0 & 0 \\
3 & 5 & 1 & 0 \\
4 & 6 & 7 & 1
\end{pmatrix}, U = \begin{pmatrix}
1 & 1 & 2 & 4 \\
0 & 3 & 4 & 1 \\
0 & 0 & 3 & 1 \\
0 & 0 & 0 & 12
\end{pmatrix}$


## Задание 3

Решить систему линейных уравнений методом $LU$-разложения

$$\begin{cases}
2x_{1}+x_{2}+3x_{3}=1 \\
11x_{1}+7x_{2}+5x_{3}=-6 \\
9x_{1}+8x_{2}+4x_{3}=-5
\end{cases}$$

### Решение

$A = 
\begin{pmatrix}
2 & 1 & 3 \\
11 & 7 & 5 \\
9 & 8 & 4 \\
\end{pmatrix}\implies
\begin{pmatrix}
1 & 3 & 2 \\
7 & 5 & 11 \\
8 & 4 & 9 \\
\end{pmatrix}
$

$1) L = \begin{pmatrix}
1 & 0 & 0 \\
7 & 1 & 0 \\
8 & l_{3,2} & 1
\end{pmatrix}, U = \begin{pmatrix}
2 &   3 &     1 \\
0 & -16 & -3 \\
0 & -20 & -7
\end{pmatrix} \\
2) L = \begin{pmatrix}
1 & 0 & 0 \\
7 & 1 & 0 \\
8 & 1.25 & 1
\end{pmatrix}, U = \begin{pmatrix}
2 &   3 &     1 \\
0 & -16 & -3 \\
0 & 0 & -3.25
\end{pmatrix}
$

$$L\cdot y = B$$



$
\begin{cases}
y_1 = 1 \\
7y_1 + y_2 = -6 \implies \\
8y_1 + 1.25 y_2 + y_3 = -5 \implies
\end{cases}\implies
\begin{cases}
y_1 = 1 \\
y_2 = -13 \\
y_3 = 3.25
\end{cases}
$

$$U\cdot x = y$$

$
\begin{cases}
2x_1 + 3x_2 + x_3 = 1 \\
-16x_2 -3x_3 = -13 \\
-3.25x_3 = 3.25
\end{cases}\implies
\begin{cases}
x_1 = -0.5 \\
x_2 = 1 \\
x_3 = -1
\end{cases}
$

$x_2$ и $x_3$ нужно поменять местами при проверке, т.к. я менял порядок столбцов для удобства

## Задание 4

Решить систему линейных уравнений методом Холецкого

$$\begin{cases}
81x_{1}-45x_{2}+45x_{3}=531 \\
-45x_{1}+50x_{2}-15x_{3}=-460 \\
45x_{1}-15x_{2}+38x_{3}=193
\end{cases}$$

### Решение

$A =
\begin{pmatrix}
81 & -45 & 45 \\
-45 & 50 & -15 \\
45 & -15 & 38
\end{pmatrix}\implies
\begin{pmatrix}
-45 & 81 & 45 \\
50 & -45 & -15 \\
-15 & 45 & 38
\end{pmatrix}
$

1) $L =
\begin{pmatrix}
1 & 0 & 0 \\
-10/9 & 1 & 0 \\
1/3 & l_{3,2} & 1
\end{pmatrix},
U = \begin{pmatrix}
-45 & 81 & 45 \\
0 & 45 & 35 \\
0 & 18 & 23
\end{pmatrix}
$

2) $L =
\begin{pmatrix}
1 & 0 & 0 \\
-10/9 & 1 & 0 \\
1/3 & 2/5 & 1
\end{pmatrix},
U = \begin{pmatrix}
-45 & 81 & 45 \\
0 & 45 & 35 \\
0 & 0 & 9
\end{pmatrix}
$

$L^T = \begin{pmatrix}
1 & -10/9 & 1/3 \\
0 & 1 & 2/5 \\
0 & 0 & 1
\end{pmatrix}$

$$Ly = b$$

$\begin{cases}
y_1 = 531 \\
-\frac{10}{9}y_1 + y_2 = -460 \\
\frac{1}{3}y_1 + \frac{2}{5}y_2 + y_3 = 193
\end{cases}\implies
\begin{cases}
y_1 = 531 \\
y_2 = 130 \\
y_3 = 292
\end{cases}$

$$L^Tx=y$$

$
\begin{cases}
x_1 - \frac{10}{9}x_2 + \frac{1}{3}x_3 = 531 \\
x_2 + \frac{2}{5}x_3 = 130 \\
x_3 = 292
\end{cases}
$

$
\begin{cases}
x_1 = 643 \\
x_2 = 13.2 \\
x_3 = 292
\end{cases}
$

## Задание 5

Написать на Python программу с реализацией одного из изученных алгоритмов решения СЛАУ.

### Решение